pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+
+	var glusterServer string
+	var bkpServers []string
+
+	if req == nil {
+		glog.Errorf("volume create request is nil")
+		return nil, status.Errorf(codes.InvalidArgument, "request cannot be empty")
+	}
+
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Name is a required field")
+	}
+	glog.V(1).Infof("creating volume with name ", req.Name)
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities is a required field")
+	}
+
+	// If capacity mentioned, pick that or use default size 20 MB
+	volSizeBytes := defaultVolumeSizeInMB
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeMB := int((volSizeBytes + MB - 1) / MB)
+
+	// Get Volume name : TODO use the values from request
+	volumeName := req.Name
+	glusterVol := req.GetParameters()["glustervol"]
+	glusterServer = req.GetParameters()["glusterserver"]
+	glusterURL := req.GetParameters()["glusterurl"]
+	glusterURLPort := req.GetParameters()["glusterurlport"]
+	glusterUser := req.GetParameters()["glusteruser"]
+	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	glog.V(2).Infof("Volume list response: %+v", volumeListResp)
+
+	/* TODO: Check the requested volume exist in the cluster
+	glusterServer, bkpServers, err = cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeBytes)
+	if err == nil {
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				Id:            volumeName,
+				CapacityBytes: int64(volSizeBytes),
+				Attributes: map[string]string{
+					"glustervol":        volumeName,
+					"glusterserver":     glusterServer,
+					"glusterbkpservers": strings.Join(bkpServers, ":"),
+				},
+			},
+		}, nil
+	}
+	*/
+
+	// If volume does not exist, provision volume
+	glog.V(4).Infof("Received request to create/provision volume name:%s with size:%d", volumeName, volSizeMB)
+	volMetaMap := make(map[string]string)
+	volMetaMap[glusterDescAnn] = glusterDescAnnValue
+	volumeReq := api.VolCreateReq{
+		Name:     volumeName,
+		Metadata: volMetaMap,
+		Size:     uint64(volSizeMB),
+	}
+
+	glog.V(2).Infof("volume request: %+v", volumeReq)
+	volumeCreateResp, err := cs.client.VolumeCreate(volumeReq)
+	if err != nil {
+		glog.Errorf("failed to create volume : %v", err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create volume: %s", err.Error()))
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
+	err = cs.client.VolumeStart(volumeName, true)
```

Comment:
I have some though on this
I agree with @kshlm, in perspective of resource utilization.

the problem here is during volume mount, what if volume start fails? kubernetes will never try to delete and recreate the volume. it will always try to mount the volume

if we start volume during the volume creation time, if volume start fails we can delete the created volume and return back an error to the kubernets, 
kubernetes will retry to create the PVC because of failure

Reply:
@kshlm @Madhu-1 By volume provisioned success, in kube it means its ready to be served. In that sense keeping it activated is the expectation. May be there can be other queries on volume independent of mount call.